File: busrep_server/src/blockchain.rs
use super::chrono::Local;
use super::crypto::digest::Digest;
use super::crypto::sha2::Sha256;
use super::models::*;

pub fn created() -> String {
    format!("{}", Local::now().naive_local())
}

pub fn previous_hash(block: Block) -> String {
    let mut hasher = Sha256::new();
    hasher.input_str(&format!(
        "{block_id}{action}{action_id}{digital_signature}{created}{previous_hash}",
        block_id = block.block_id,
        action = block.action,
        action_id = block.action_id,
        digital_signature = block.digital_signature,
        created = block.created,
        previous_hash = block.previous_hash
    ));
    hasher.result_str()
}

File: busrep/lib/helpers/blockchain.dart
import 'dart:convert';

import 'package:busrep/helpers/crypto.dart';
import 'package:busrep/models/blockchain.dart';
import 'package:busrep/repositories/database.dart';

Future<int> loadMaxBlockID() async {
  Block lastBlock = await getLastBlock();
  return lastBlock.blockID;
}

void blockchained(Blockchain responsePost) async {
  if (await isLegalBlockChain(responsePost)) {
    saveBlockchain(responsePost);
  } else {
    throw Exception('Error: Blockchain is Illegal !!!');
  }
}

Future<bool> isLegalBlockChain(Blockchain responsePost) async {
  Block previousBlock = await getLastBlock();
  for (Block block in responsePost.blockchain) {
    if ((await previousHash(previousBlock)) != block.previousHash) {
      return false;
    }
    previousBlock = block;
  }
  return true;
}

Future<String> previousHash(Block previousBlock) async {
  String message = previousBlock.blockID.toString() +
      previousBlock.action +
      previousBlock.actionID.toString() +
      jsonEncode(previousBlock.digitalSignature) +
      previousBlock.created +
      previousBlock.previousHash;
  List<int> hashedMessage = await hashMessage(message);
  var previousHash = "";
  hashedMessage.forEach((c) {
    String c2 = c.toRadixString(16);
    if (c2.length == 1) {
      c2 = "0" + c2;
    }
    previousHash += c2;
  });
  return previousHash;
}
